### PR TITLE
feat: paths as objects + user-provided Media cls + handle static

### PR DESCRIPTION
**Pull request by JuroOravec**
Changes:
- Allow to specify template/js/css files as str, bytes, FileLike, SafeString, or a function that returns one of the other types.
- Support Django's "path as objects" feature - when an object is a "safe string" (defines `__html__` method), then we do NOT try to resolve it.
- Allow to override how JS/CSS paths are rendered into `<script>`/`<link>` tags by subclassing Django's `Media` class and supplying that to `Component.media_class`
- ~Validate that keys are valid CSS media types when `Media.css` is defined as a dictionary~

Documents the above plus also how component's template/css/js paths are resolved relative to component's dir if possible.

While we should wait to see what the guys from https://github.com/EmilStenstrom/django-components/issues/522 say, the tests suggest that this should fix their issue.

Closes https://github.com/EmilStenstrom/django-components/issues/522

### CODE REVIEW

File: src/django_components/component.py [hunk 1]
```diff
@@ -54,37 +55,113 @@
...
+        # `MediaDefiningClass` defines `media` as a getter (defined in djago.forms.widgets.media_property).
+        # So we reused that and convert it to user-defined Media class
+        media_prop: property = cls.media
+        media_getter = media_prop.fget
+        def media_wrapper(self: "Component") -> Any:
+            if not media_getter:
+                return None
+            media: Media = media_getter(self)
+            return self.media_class(js=media._js, css=media._css)
+        cls.media = property(media_wrapper)
```
Comment (JuroOravec) [inline src/django_components/component.py:86]: The bit above makes it possible for users to set `media_class` on the component class to override the Media class. For examples see https://github.com/EmilStenstrom/django-components/pull/526/files#diff-7667a514b72a24c770553fe3421223dcc03252c458518310f25a5ef7746d4669R473 and https://github.com/EmilStenstrom/django-components/pull/526/files#diff-7667a514b72a24c770553fe3421223dcc03252c458518310f25a5ef7746d4669R496

File: src/django_components/component.py [hunk 2]
```diff
@@ -54,37 +55,113 @@
...
             return super().__new__(mcs, name, bases, attrs)
 
         if "Media" in attrs:
-            media: Component.Media = attrs["Media"]
+            media_data: Component.Media = attrs["Media"]
+            _normalize_media(media_data)
+            _normalize_media_filepaths(media_data)
 
-            # Allow: class Media: css = "style.css"
-            if hasattr(media, "css") and isinstance(media.css, str):
```
Comment (JuroOravec) [inline src/django_components/component.py:68]: Moved the normalization to a function for clarity

File: src/django_components/component.py [hunk 3]
```diff
@@ -54,37 +55,113 @@
...
         # NOTE: Skip template/media file resolution when then Component class ITSELF
         # is being created.
         if "__module__" in attrs and attrs["__module__"] == "django_components.component":
             return super().__new__(mcs, name, bases, attrs)
 
         if "Media" in attrs:
-            media: Component.Media = attrs["Media"]
+            media_data: Component.Media = attrs["Media"]
+            _normalize_media(media_data)
+            _normalize_media_filepaths(media_data)
```
Comment (JuroOravec) [inline src/django_components/component.py:68]: `_normalize_media_filepaths` what allows users to define file paths as str/bytes/PathLike/Callable

File: src/django_components/component.py [hunk 4]
```diff
@@ -54,37 +55,113 @@
...
+        return True
+    
+    return False
+
+
+def _normalize_media_filepath(filepath: Any) -> str | SafeData:
+    print("_normalize_media_filepath: ", filepath)
+
+    if isinstance(filepath, SafeData) or hasattr(filepath, "__html__"):
+        return filepath
```
Comment (JuroOravec) [inline src/django_components/component.py:158]: This bit above ensures that when users creates an "object path" with `@ html_safe`, then we don't try to convert it to a string.

File: src/django_components/component.py [hunk 5]
```diff
@@ -126,15 +203,17 @@ def _resolve_component_relative_files(attrs: MutableMapping) -> None:
...
-        maybe_resolved_filepath = os.path.join(comp_dir_abs, filepath)
-        component_import_filepath = os.path.join(comp_dir_rel, filepath)
-
-        if os.path.isfile(maybe_resolved_filepath):
-            logger.debug(
-                f"Interpreting template '{filepath}' of component '{module_name}' relatively to component file"
-            )
-            return component_import_filepath
+    def resolve_file(filepath: str | SafeData) -> str | SafeData:
+        if isinstance(filepath, str):
```
Comment (JuroOravec) [inline src/django_components/component.py:212]: Path resolution logic is applied only to paths that resolved to plain string (so paths defined as str, bytes, PathLike, functions).

File: src/django_components/component.py [hunk 6]
```diff
@@ -126,15 +203,17 @@ def _resolve_component_relative_files(attrs: MutableMapping) -> None:
...
+            maybe_resolved_filepath = os.path.join(comp_dir_abs, filepath)
+            component_import_filepath = os.path.join(comp_dir_rel, filepath)
+
+            if os.path.isfile(maybe_resolved_filepath):
+                logger.debug(
+                    f"Interpreting template '{filepath}' of component '{module_name}' relatively to component file"
+                )
+                return static(component_import_filepath)
+            return static(filepath)
```
Comment (JuroOravec) [inline src/django_components/component.py:221]: And here we apply the `static()` call to the string path which has been possibly resolved as a relative path,

File: tests/test_component_media.py [hunk 7]
```diff
@@ -282,6 +286,250 @@ class Media:
             """,
         )
 
+    # TODO - ALLOW CALLABLE!
+    # TODO - DOCUMENT IT ALL!
+    # TODO - Document how when we pass a safe string to css/js, then
+    #        Media.render_js/css DOES NOT format it (same as Django's
+    #        see https://docs.djangoproject.com/en/5.0/topics/forms/media/#paths-as-objects)
```
Comment (EmilStenstrom) [inline tests/test_component_media.py:293]: TODO

File: tests/test_component_media.py [hunk 8]
```diff
@@ -282,6 +286,250 @@ class Media:
...
+            <script src="path/to/script2.js"></script>
+            """,
+        )
+
+    def test_media_custom_render_js(self):
+        class MyMedia(Media):
+            def render_js(self):
+                tags: list[str] = []
+                for path in self._js:
+                    tags.append(f'<my_script_tag src="{self.absolute_path(path)}"></my_script_tag>')
```
Comment (EmilStenstrom) [inline tests/test_component_media.py:484]: `absolute_path(path)` doesn't seem to do anything?

**Review by EmilStenstrom — COMMENTED** [review]
Really like the API here! Looking forward to seeing the final PR!

File: tests/test_component_media.py (hunk 8 above)
Comment (JuroOravec) [reply tests/test_component_media.py:588]: In that tests, `absolute_path(path)` did nothing, because neither `staticfiles` app was given, nor was `STATIC_URL` set. I've added a separate test here https://github.com/EmilStenstrom/django-components/pull/526/files#diff-7667a514b72a24c770553fe3421223dcc03252c458518310f25a5ef7746d4669R530 to the use of `static(path)` / `absolute_path(path)` when `STATIC_URL` is given.

File: src/django_components/component.py [hunk 9]
```diff
@@ -49,145 +48,22 @@
...
-            # Allow: class Media: js = "script.js"
-            if hasattr(media, "js") and isinstance(media.js, str):
-                media.js = [media.js]
-
-        _resolve_component_relative_files(attrs)
-
         return super().__new__(mcs, name, bases, attrs)
 
 
-def _resolve_component_relative_files(attrs: MutableMapping) -> None:
```
Comment (JuroOravec) [inline src/django_components/component.py:90]: I've moved the logic that prepares the Media class to `component_media`

File: README.md [hunk 10]
```diff
@@ -1645,7 +1650,211 @@ If you find yourself using the `only` modifier often, you can set the [context_b
 
 Components can also access the outer context in their context methods like `get_context_data` by accessing the property `self.outer_context`.
 
-## Rendering JS and CSS dependencies
+## Defining HTML/JS/CSS files
```
Comment (JuroOravec) [inline README.md:1653]: Documented how to work with the Media class, the inputs, how to override it, etc. So conceptually there's 2 separate steps:
1. "Defining HTML/JS/CSS files" describes how to define/link files to a components
2. "Rendering JS and CSS dependencies" describes how, given that I have components that have linked files, how I render out the `<script>`/`<link>` tags.

File: src/django_components/component_media.py [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+from django_components.utils import search
+
+if TYPE_CHECKING:
+    from django_components.component import Component
+
+
+# For full list, see https://docs.djangoproject.com/en/5.0/topics/forms/media/#css
+# Note that many of these are marked as deprecated. For the recommended
+# media types, see https://developer.mozilla.org/en-US/docs/Web/CSS/@media#media_types.
+CSS_MEDIA_TYPES = (
```
Comment (JuroOravec) [inline src/django_components/component_media.py:19]: Normally Django's Media class accepts a dictionary for the `css` attribute. There, the dict keys represent [CSS media types](https://developer.mozilla.org/en-US/docs/Web/CSS/@media#media_types). There is only a small set of all the media types, so I added validation

File: src/django_components/component_media.py [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+
+
+def _resolve_component_relative_files(attrs: MutableMapping) -> None:
+    """
+    Check if component's HTML, JS and CSS files refer to files in the same directory
+    as the component class. If so, modify the attributes so the class Django's rendering
+    will pick up these files correctly.
+    """
+    # First check if we even need to resolve anything. If the class doesn't define any
+    # JS/CSS files, just skip.
```
Comment (JuroOravec) [inline src/django_components/component_media.py:277]: `_resolve_component_relative_files` and `_get_dir_path_from_component_path` didn't change, just added the check here to skip the relative path resolution logic when the component class doesn't define any files.

File: tests/test_component_media.py [hunk 13]
```diff
@@ -282,6 +287,437 @@ class Media:
...
+
+            <script hi src="calendar/script.js"></script>
+            <script src="calendar/script1.js"></script>
+            <script src="calendar/script2.js"></script>
+            <script src="calendar/script3.js"></script>
+            """,
+        )
+
+    @override_settings(STATIC_URL="static/")
+    def test_works_with_static(self):
```
Comment (JuroOravec) [inline tests/test_component_media.py:579]: Added a separate test here to check that the paths defined ins `Media.css` / `Media.js` work with staticfiles.

File: tests/test_component_media.py [hunk 14]
```diff
@@ -282,6 +287,437 @@ class Media:
...
+            <script hi src="/static/calendar/script.js"></script>
+            <script src="/static/calendar/script1.js"></script>
+            <script src="/static/calendar/script2.js"></script>
+            <script src="/static/calendar/script3.js"></script>
+            <script src="/static/calendar/script4.js"></script>
+            """,
+        )
+
+
+class MediaStaticfilesTests(BaseTestCase):
```
Comment (JuroOravec) [inline tests/test_component_media.py:618]: These following 2 tests use a similar setup as described in https://github.com/EmilStenstrom/django-components/issues/522. The first test uses a default Storage class, while second uses `ManifestStaticFilesStorage`.

**Comment by JuroOravec** [comment]
@EmilStenstrom This one is now ready for review :)

File: src/django_components/component_media.py [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+
+def _normalize_media(media: ComponentMediaInput) -> None:
+    if hasattr(media, "css") and media.css:
+        # Allow: class Media: css = "style.css"
+        if _is_media_filepath(media.css):
+            media.css = [media.css]  # type: ignore[list-item]
+
+        # Allow: class Media: css = ["style.css"]
+        if isinstance(media.css, (list, tuple)):
+            media.css = {"all": media.css}
```
Comment (EmilStenstrom) [inline src/django_components/component_media.py:191]: Are we changing the media type from "None" to "all" here? media="all" is implicit in CSS so I don't think we should add that to everyone's CSS when it's not needed.

File: tests/components/relative_file/relative_file_pathobj.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+    def post(self, request, *args, **kwargs) -> HttpResponse:
+        variable = request.POST.get("variable")
+        return self.render_to_response({"variable": variable})
+
+    def get(self, request, *args, **kwargs) -> HttpResponse:
+        return self.render_to_response({"variable": "GET"})
+
+    def get_context_data(self, variable, *args, **kwargs) -> Dict[str, Any]:
+        return {"variable": variable}
```
Comment (EmilStenstrom) [inline tests/components/relative_file/relative_file_pathobj.py:41]: Not a super fan of having shared fixtures used for testing like this, because it makes tests hard to follow (you have to open several files to understand them), and can introduce cross-test dependencies inadvertently. Not super important, because I think we're already doing this in other places in the tests.

File: tests/test_autodiscover.py [hunk 17]
```diff
@@ -36,7 +36,7 @@ def test_autodiscover_with_components_as_views(self):
 
         all_components_after = component_registry.registry.all().copy()
         imported_components_count = len(all_components_after) - len(all_components_before)
-        self.assertEqual(imported_components_count, 1)
+        self.assertEqual(imported_components_count, 2)
```
Comment (EmilStenstrom) [inline tests/test_autodiscover.py:39]: This is a very strange test. Checks that we didn't add any new autodiscoverable components?

File: tests/test_component_media.py [hunk 18]
```diff
@@ -282,6 +287,448 @@ class Media:
...
+        # Configure static files. The dummy files are set up in the `./static_root` dir.
+        # The URL should have path prefix /static/.
+        # NOTE: We don't need STATICFILES_DIRS, because we don't run collectstatic
+        #       See https://docs.djangoproject.com/en/5.0/ref/settings/#std-setting-STATICFILES_DIRS
+        STATIC_URL="static/",
+        STATIC_ROOT=os.path.join(Path(__file__).resolve().parent, "static_root"),
+        # NOTE: STATICFILES_STORAGE is deprecated since 5.1, use STORAGES instead
+        #       See https://docs.djangoproject.com/en/5.0/ref/settings/#staticfiles-storage
+        STORAGES={
+            # This was NOT changes
```
Comment (EmilStenstrom) [inline tests/test_component_media.py:686]:
```suggestion
            # This was NOT changed
```

File: tests/test_component_media.py [hunk 19]
```diff
@@ -282,6 +287,448 @@ class Media:
...
+
+        comp = SimpleComponent()
+
+        # NOTE: Since we're using ManifestStaticFilesStorage, we expect the rendered media to link
+        # to the files as defined in staticfiles.json
+        self.assertHTMLEqual(
+            comp.render_dependencies(),
+            """
+            <link href="/static/calendar/style.0eeb72042b59.css" media="all" rel="stylesheet">
+            <my_script_tag src="/static/calendar/script.e1815e23e0ec.js"></my_script_tag>
```
Comment (EmilStenstrom) [inline tests/test_component_media.py:728]: Really nice that we can support this now!

**Review by EmilStenstrom — APPROVED** [review]
Really nice work as always @JuroOravec!

Added some small nitpicks, and a question about if we can render media="all" in fewer cases. In general I approve! :)

File: src/django_components/component_media.py (hunk 15 above)
Comment (JuroOravec) [reply src/django_components/component_media.py:191]: I didn't implement it, so I might be missing something, but I'm not aware of the "None" option, was that maybe something in older Django versions? Because in v5, when you [go to the docs](https://docs.djangoproject.com/en/5.0/topics/forms/media/#css) and see the examples there, it looks like `"all"` is the default / catch-all option.

File: tests/components/relative_file/relative_file_pathobj.py (hunk 16 above)
Comment (JuroOravec) [reply tests/components/relative_file/relative_file_pathobj.py]: That's a good point, I've moved this file to it's own dir `relative_file_pathobj`

File: tests/test_autodiscover.py (hunk 17 above)
Comment (JuroOravec) [reply tests/test_autodiscover.py:41]: Hm, looking at the test, I think the test would make more sense if we checked that `autoimport()` imports the components from the `tests/components` dir.

For now I've put a TODO there. I will be reviewing these tests when I'll be making an MR for autoimport, as mentioned here https://github.com/EmilStenstrom/django-components/issues/527#issuecomment-2183091056

File: src/django_components/component_media.py (hunk 11 above)
Comment (JuroOravec) [reply src/django_components/component_media.py:19]: Aha, I made a mistake. In Django docs, they use also this example, with two values separated by comma:

```py
class Media:
    css = {
        "tv,projector": ["lo_res.css"],
    }
```
Which corresponds to:
```html
<link href="http://static.example.com/lo_res.css" media="tv,projector" rel="stylesheet">
```

So I assume that this means that they keys in the dictionary can use also the [logical operators](https://developer.mozilla.org/en-US/docs/Web/CSS/@media#logical_operators) (comma, `not`, `only`, etc...).

So in that case IMO validating the CSS keys is not worth it, as there could be many different combinations.

File: src/django_components/component_media.py (hunk 15 above)
Comment (JuroOravec) [reply src/django_components/component_media.py:173]: Btw @EmilStenstrom I'll merge this MR so we can close the original issue. But if there's a way to not generate the `media="all"`, e.g. with the "None" option, we can open a new issue for that or I can just include it in on of my other MRs

Comment (EmilStenstrom) [reply src/django_components/component_media.py:173]: I was thinking that `css = "style.css"` (our version) would generate a link tag without media set. But let's keep it the way to wrote it.
